File: problem/0098/validate_binary_search_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidSub(TreeNode* node, TreeNode* min=NULL, TreeNode* max=NULL) {
        if (node == NULL)
            return true;
        
        if (min && node->val <= min->val)
            return false;
        if (max && node->val >= max->val)
            return false;
        if (!isValidSub(node->left, min, node) || !isValidSub(node->right, node, max))
            return false;
        
        return true;
    }
    
    bool isValidBST(TreeNode* root) {
        return isValidSub(root, NULL, NULL);;
    }
};

File: problems/0098/ts/validate_binary_search_tree.ts
import { TreeNode } from '../../../structures/ts/tree';

const isValidSub = (
  root: TreeNode | null,
  min: number,
  max: number
): boolean => {
  if (root == null) return true;

  if (root.val >= max) return false;
  if (root.val <= min) return false;

  return (
    isValidSub(root.left, min, root.val) &&
    isValidSub(root.right, root.val, max)
  );
};

const isValidBST = (root: TreeNode | null): boolean => {
  if (root == null) return true;
  return isValidSub(root, Number.MIN_SAFE_INTEGER, Number.MAX_SAFE_INTEGER);
};

export default isValidBST;
